packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -0,0 +1,125 @@
+  public readonly service = 'custom-s3-deployment';
+
+  constructor(private readonly functionName: string, private readonly customResourceProperties: any) {
+  }
+
+  public async apply(sdk: ISDK): Promise<any> {
+    // JSON.stringify() doesn't turn the actual objects to strings, but the lambda expects strings
+    const resourceProperties = stringifyObject(this.customResourceProperties);
+
+    return sdk.lambda().invoke({
+      FunctionName: this.functionName,
+      // Lambda refuses to take a direct JSON object and requires it to be stringify()'d
+      Payload: JSON.stringify({
+        RequestType: 'Update',
+        ResponseURL: REQUIRED_BY_CFN,
+        PhysicalResourceId: REQUIRED_BY_CFN,
+        StackId: REQUIRED_BY_CFN,
+        RequestId: REQUIRED_BY_CFN,
+        LogicalResourceId: REQUIRED_BY_CFN,
+        ResourceProperties: resourceProperties,
+      }),
+    }).promise();
+  }
+}
+
+async function changeIsForS3DeployCustomResourcePolicy(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapImpact | EmptyHotswapOperation> {
+  const roles = change.newValue.Properties?.Roles;
+  if (!roles) {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  for (const role of roles) {
+    const roleLogicalId = await evaluateCfnTemplate.findLogicalIdForPhysicalName(await evaluateCfnTemplate.evaluateCfnExpression(role));
+    if (!roleLogicalId) {
+      return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+    }
+
+    const roleRefs = evaluateCfnTemplate.findReferencesTo(roleLogicalId);
+    for (const roleRef of roleRefs) {
+      if (roleRef.Type === 'AWS::Lambda::Function') {
+        const lambdaRefs = evaluateCfnTemplate.findReferencesTo(roleRef.LogicalId);
+        for (const lambdaRef of lambdaRefs) {
+          // If S3Deployment -> Lambda -> Role and IAM::Policy -> Role, then this IAM::Policy change is an
+          // artifact of old-style synthesis
+          if (lambdaRef.Type !== 'Custom::CDKBucketDeployment') {
+            return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+          }
+        }
+      } else if (roleRef.Type === 'AWS::IAM::Policy') {
+        if (roleRef.LogicalId !== logicalId) {
+          return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+        }
+      } else if (roleRef.Type !== 'AWS::IAM::Policy') {
+        return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+      }
+    }
+  }
+
+  return new EmptyHotswapOperation();
+}
+
+function stringifyObject(obj: any): any {
+  if (obj == null) {
+    return obj;
+  }
+  if (Array.isArray(obj)) {
+    return obj.map(stringifyObject);
+  }
+  if (typeof obj !== 'object') {
+    return obj.toString();
+  }
+
+  const ret: { [k: string]: any } = {};
+  for (const [k, v] of Object.entries(obj)) {
+    ret[k] = stringifyObject(v);
+  }
+  return ret;
+}
+
+class EmptyHotswapOperation implements HotswapOperation {
+  readonly service = 'AWS::IAM::Policy';
```

Comment:
I don't think this is correct, or in line with the existing conventions for this field.

I see two options here:

1. Either make this field optional in `HotswapOperation`, and just don't set it in the code that uses it when it's `undefined`;
2. Or, set this to `'empty'`.

Your call.